pyquil/api/_qpu_client.py

```diff
@@ -0,0 +1,169 @@
+##############################################################################
+# Copyright 2016-2021 Rigetti Computing
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+##############################################################################
+from contextlib import contextmanager
+from dataclasses import dataclass
+from typing import Dict, Any, Iterator, cast, Tuple, Union, List
+
+import rpcq
+from qcs_api_client.models import EngagementWithCredentials, EngagementCredentials
+
+
+@dataclass
+class RunProgramRequest:
+    """
+    Request to run a program.
+    """
+
+    id: str
+    """Identifier for request."""
+
+    priority: int
+    """Priority for request."""
+
+    program: str
+    """Encrypted program to run."""
+
+    patch_values: Dict[str, List[Union[int, float]]]
+    """Map of data names to data values for patching the program."""
+
+
+@dataclass
+class RunProgramResponse:
+    """
+    Run program response.
+    """
+
+    job_id: str
+    """Identifier for created run job."""
+
+
+@dataclass
+class GetBuffersRequest:
+    """
+    Request for getting job buffers.
+    """
+
+    job_id: str
+    """Job for which to get buffers."""
+
+    wait: bool
+    """Whether or not to wait until buffers become available."""
+
+
+@dataclass
+class BufferResponse:
+    """
+    Job buffer.
+    """
+
+    shape: Tuple[int, int]
+    """Shape of the buffer: (<trials>, <slots>)."""
+
+    dtype: str
+    """Buffer data type."""
+
+    data: bytes
+    """Raw buffer data (C order)."""
+
+
+@dataclass
+class GetBuffersResponse:
+    """
+    Job buffers response.
+    """
+
+    buffers: Dict[str, BufferResponse]
+    """Job buffers, by buffer name."""
+
+
+class QPUClient:
+    """
+    Client for making requests to a QPU.
+    """
+
+    def __init__(self, *, engagement: EngagementWithCredentials, request_timeout: float = 5.0) -> None:
```

Comment:
Earlier in this PR I made a comment about the possibility of this engagement going stale during the lifetime of a QPUClient instance. The client is intended to be short-lived (hence the with client_provider.qpu_client(...) contextmanager usage), but still something to think about. Also, being short-lived doesn't necessarily protect from the problem. For every duration of this client, there's always the remote chance that you end up with an engagement that has a shorter remaining lifetime.

One possibility is to have this client take an EngagementManager instead of an EngagementWithCredentials, but I don't love that idea. If it came to that, we might as well merge the EngagementManager functionality into this class itself (a potentially less bad idea, but makes this client more stateful and less "functional-programmy").

Reply:
A third idea is to make the `EngagementManager` refresh engagements a few minutes early (or some amount of time we deem appropriate). That way, the engagement will have some guaranteed minimum amount of lifetime left.

We could also implement this on the server side, allowing engagements to be used a few minutes after expiration for a transition period, instead of having to implement it for each client. I think this is similar to what 2FA tokens do. When a token expires, I think there's a brief window of time where the token will still work AND the new token will work (that's based on my own anecdotal evidence -- not sure if this is actually 100% true).

Reply:
The benefit to the server-side approach is that we could defer this and wait to see if it's actually a problem for users, then fix it without a need for users to update pyquil.

Reply:
I prefer that anticipating expiration in the client and refreshing pre-emptively (that's what we do with JWTs in many cases). I think we'll want a hard expiration on the server to enforce reservation boundaries.

Another thought would define an abstract class or union type between `EngagementWithCredentials` and `EngagementManager`. In the latter case, the calling code could use the manager to refresh as appropriate and fail in the former (hopefully avoiding any esoteric hanging in the case of expired credentials).

Reply:
Ah yeah, enforcing engagement boundaries makes sense. Since it sounds like we have a pattern with JWTs, maybe I'll go with refreshing early then. How long ahead of time would you say?

Reply:
I think the retry functionality could work well. I'll give it a try (probably in the `QPU` class where the engagement is fetched and the `QPUClient` is used to make the request). Unless we think the retry logic should be a lower-level (`QPUClient`) concern. That might take a bit more refactoring (e.g. this client would need to take in an `EngagementManager` rather than an engagement), but it could be done.

Reply:
If we did go the route of having this client take in an `EngagementManager`, should we just merge its functionality into this client? It's really the only client that needs engagements.